### PR TITLE
Add CSV format for trades

### PR DESCRIPTION
**Pull request by jodastephen**
FRA, Swap and Term Deposit.

All can be created from a convention or from individual fields.
Swaps do not support variable notional/rate/gearing/spread, but support almost all other features.

### CODE REVIEW

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+        FixedRateStubCalculation.ofKnownAmount(CurrencyAmount.of(currency, v))));
+    // final stub
+    Optional<Double> finalStubRateOpt = findValue(row, leg, FINAL_STUB_RATE_FIELD)
+        .map(s -> TradeCsvLoader.parseDoublePercent(s));
+    Optional<Double> finalStubAmountOpt = findValue(row, leg, FINAL_STUB_AMOUNT_FIELD)
+        .map(s -> TradeCsvLoader.parseDouble(s));
+    if (finalStubRateOpt.isPresent() && finalStubAmountOpt.isPresent()) {
+      throw new IllegalArgumentException(
+          "Swap leg must not define both '" + FINAL_STUB_RATE_FIELD + "' and  '" + FINAL_STUB_AMOUNT_FIELD + "'");
+    }
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java]: Tried, but seems more confusing.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/SwapTradeCsvLoader.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+   * 
+   * @param row  the CSV row
+   * @param info  the trade info
+   * @param refData  the reference data
+   * @return the loaded trades, all errors are captured in the result
+   */
+  static SwapTrade parse(CsvRow row, TradeInfo info, ReferenceData refData) {
+    Optional<String> conventionOpt = row.findValue(CONVENTION_FIELD);
+    if (conventionOpt.isPresent()) {
+      // using an 'unnecessary' nested class to allow parsing logic to be split
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/SwapTradeCsvLoader.java]: It got removed

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/TradeCsvLoader.java [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+        return LocalDate.parse(str, YYYYMMDD);
+      }
+      // dMMMyyyy (and all others)
+      return LocalDate.parse(str, D_MMM_YYYY_NODASH);
+
+    } catch (DateTimeParseException ex) {
+      throw new IllegalArgumentException(
+          "Unknown date format, must be formatted as " +
+              "yyyy-MM-dd, yyyyMMdd, dd/MM/yyyy, yyyy/MM/dd, 'd-MMM-yyyy' or 'dMMMyyyy' but was: " + str,
+          ex);
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/TradeCsvLoader.java]: Exceptions are only including the recommended inputs, not everything we accept ;-)

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+  private static final String FIXING_OFFSET_CAL_FIELD = "Fixing Offset Calendar";
+  private static final String FIXING_OFFSET_ADJ_CNV_FIELD = "Fixing Offset Adjustment Convention";
+  private static final String FIXING_OFFSET_ADJ_CAL_FIELD = "Fixing Offset Adjustment Calendar";
+  private static final String NEGATIVE_RATE_METHOD_FIELD = "Negative Rate Method";
+  private static final String FIRST_RATE_FIELD = "First Rate";
+  private static final String ACCRUAL_METHOD_FIELD = "Accrual Method";
+  private static final String RATE_CUT_OFF_DAYS_FIELD = "Rate Cut Off Days";
+  private static final String INFLATION_LAG_FIELD = "Inflation Lag";
+  private static final String INFLATION_METHOD_FIELD = "Inflation Method";
+  private static final String INFLATION_FIRST_VALUE_FIELD = "Inflation First Value";
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java]: My feeling is to stick with the names used in Strata. "Initial Index Level" is the name in FpML FYI.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+    }
+    throw new IllegalArgumentException("Swap trade index not known: " + indexStr);
+  }
+
+  //-------------------------------------------------------------------------
+  // fixed rate calculation
+  private static RateCalculation parseFixedRateCalculation(CsvRow row, String leg, double fixedRate, Currency currency) {
+    FixedRateCalculation.Builder builder = FixedRateCalculation.builder();
+    // basics
+    builder.dayCount(DayCount.of(getValue(row, leg, DAY_COUNT_FIELD)));
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java]: This needs to be tackled separately

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+      String daysField,
+      String daysCalField,
+      String cnvField,
+      String calField) {
+
+    Optional<Integer> daysOpt = findValue(row, leg, daysField)
+        .map(s -> new Integer(s));
+    HolidayCalendarId cal = findValue(row, leg, daysCalField)
+        .map(s -> HolidayCalendarId.of(s))
+        .orElse(HolidayCalendarIds.NO_HOLIDAYS);
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java]: This needs to be tackled separately

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java (hunk 4 above)
Comment (brianweller89) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/FullSwapTradeCsvLoader.java]: Fine to leave it but my main concern was that the name is a bit ambiguous to user constructing  csv file; does it refer to the initial index level or the the actual rate fixing of the first inflation period.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/SwapTradeCsvLoader.java [hunk 7]
```diff
@@ -106,16 +106,24 @@ static SwapTrade parseWithConvention(CsvRow row, TradeInfo info, ReferenceData r
...
+            "Swap trade had invalid combination of fields. When these fields are found " +
+                ImmutableList.of(CONVENTION_FIELD, START_DATE_FIELD, TENOR_FIELD) +
+                " then these fields must not be present " +
+                ImmutableList.of(PERIOD_TO_START_FIELD, END_DATE_FIELD));
       }
+      LocalDate startDate = startDateOpt.get();
+      Tenor tenor = tenorOpt.get();
+      LocalDate endDate = startDate.plus(tenor);
+      SwapTrade trade = createSwap(info, conventionStr, startDate, endDate, buySell, notional, fixedRate, fxRateOpt);
+      return adjustTrade(trade, rollCnvOpt, stubCnvOpt, firstRegStartDateOpt, lastRegEndDateOpt, dateCnv, dateCalOpt);
```
Comment (brianweller89) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/SwapTradeCsvLoader.java]: Can we update the javadoc (TradeCsvLoader) to reflect the fact that this combination is now supported?

**Pull request merged**